flink-clients/src/main/java/org/apache/flink/client/deployment/application/ApplicationMiniCluster.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.client.deployment.application;
+
+import org.apache.flink.api.common.JobSubmissionResult;
+import org.apache.flink.client.program.PackagedProgram;
+import org.apache.flink.runtime.dispatcher.SessionDispatcherFactory;
+import org.apache.flink.runtime.dispatcher.runner.DefaultDispatcherRunnerFactory;
+import org.apache.flink.runtime.entrypoint.component.DefaultDispatcherResourceManagerComponentFactory;
+import org.apache.flink.runtime.entrypoint.component.DispatcherResourceManagerComponentFactory;
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.runtime.resourcemanager.StandaloneResourceManagerFactory;
+import org.apache.flink.runtime.rest.JobRestEndpointFactory;
+
+import java.util.concurrent.CompletableFuture;
+
+/** An extension to {@link MiniCluster} that runs packaged program in application mode. */
+public class ApplicationMiniCluster extends MiniCluster {
```

Comment:
Agreed, but this is hard to do without breaking mini cluster public API 🤔

I did attempt with AbstractMiniCluster and it didn't look right. 😢 Huge change-set that is breaking public API (eg. enum for HAService would have to move to abstract class, configurations, ...). I think this could be done properly, but it would definitely take some time (~ 2MD?), there should be an easier way.

What I can do is to allow passing custom DispatcherResourceManagerComponentFactory into `TestingMiniCluster` via. User would be still able to access methods like `submitJob`, but that would be no different from real world scenario, where this methods are accessible via RPC.